### DISCUSSION TITLE
How do i get the tabs to view what process is running?

### DISCUSSION THREAD

**Discussion by McFrappe** [wezterm/wezterm#1421] [Q&A] [answered]
I am trying to get the tabs to show for instance that i am running "nvim" or any other process. So, for when nothing is running on a terminal window, it should preview probably zsh since it is the underlying process for the terminal window.

Should i use the functionalities of wezterm's [mux-process](https://wezfurlong.org/wezterm/config/lua/mux-events/mux-is-process-stateful.html?highlight=process#mux-is-process-stateful)?

Currently my config looks as follows:

```lua
local wezterm = require 'wezterm';
local COLOR_BG = "white"
local COLOR_BG_ALT = "#485460"
local COLOR_FG = "white"
local COLOR_FG_ALT = "#ff5e57"
local COLOR_ACCENT = "#1e272e"

local function strip_home_name(text)
    local username = os.getenv("USER")
    return text:gsub("/home/" .. username, "~")
end


wezterm.on("update-right-status", function(window)
    local window_dimensions = window:get_dimensions();

    if not window_dimensions.is_full_screen then
        window:set_right_status(wezterm.format({{Text=""},}));
        return;
    else
        local date = wezterm.strftime("%a %b %-d %H:%M:%S");
        local bat = ""
        for _, b in ipairs(wezterm.battery_info()) do
            bat = string.format("%.0f%%", b.state_of_charge * 100)
        end

        window:set_right_status(wezterm.format({
            {Text=bat .. "   " .. date},
        }));

        return;
    end
end)

wezterm.on("format-window-title", function(tab, tabs)
	local zoomed = ""
	if tab.active_pane.is_zoomed then
		zoomed = "[Z] "
	end

	local index = ""
	if #tabs > 1 then
		index = string.format("[%d/%d] ", tab.tab_index + 1, #tabs)
	end

	local clean_title = strip_home_name(tab.active_pane.title)
	return zoomed .. index .. clean_title
end)

wezterm.on("format-tab-title", function(tab, hover)
	-- edge icon
	-- local edge_background = COLOR_BG
	-- inactive tab
	local background = COLOR_BG_ALT
	local foreground = COLOR_FG

	if tab.is_active then
		background = COLOR_FG_ALT
		foreground = COLOR_BG
	elseif hover then
		background = COLOR_ACCENT
		foreground = COLOR_FG
	end

	-- local edge_foreground = background
    local clean_title = strip_home_name(tab.active_pane.title)

	return {
		-- { Background = { Color = edge_background } },
		-- { Foreground = { Color = edge_foreground } },
		{ Background = { Color = background } },
		{ Foreground = { Color = foreground } },
		{ Text = clean_title },
		-- { Background = { Color = edge_background } },
		-- { Foreground = { Color = edge_foreground } },
	}
end)

return {
    enable_tab_bar = true,
    adjust_window_size_when_changing_font_size = false,
    window_background_opacity = 0.9,
    color_scheme = "Jellybeans",
    font = wezterm.font("JetBrains Mono", {weight = "Bold", italic = true}),
    launch_menu = {
        {
            args = {"htop"},
        },
        {
            label = "Bash Shell",
            args = {"bash"},
        },
    },
}
```

Thankful for any guidance on how to achieve this.

**Comment by wez** [comment, marked as answer]
https://github.com/wez/wezterm/issues/876 is asking for a way to pull process information for a pane.
I've been resisting this because it assumes that panes are local, and with the multiplexer functionality that isn't guaranteed.

What I use for this sort of thing is configure the `preexec` function in my shell (I use `zsh`) so that it emits escape sequences that change the tab title just prior to running commands as I type them in to the shell.  Since I use this on most of the systems I use, it also works for me when I log into remote systems.

Here's an excerpt from my dotfiles; this isn't the full thing because I also have a variation on the [shell-integration](https://github.com/wez/wezterm/blob/main/assets/shell-integration/wezterm.sh) that is [mentioned in these docs](https://wezfurlong.org/wezterm/shell-integration.html) in my actual dotfiles, and that is a bit much to recreate here:

```zsh
function _dotfiles_ansi_dequote()
{
  # `strip-ansi-escapes` is a utility that ships with wezterm and filters out ansi escape sequences
  if hash strip-ansi-escapes 2>/dev/null ; then
    echo "$1" | strip-ansi-escapes
  else
    echo "$1" | perl -pe "s/\\x1b\[[0-9]+(;[0-9]+)*[km]//g"
  fi
}

function _dotfiles_set_window_title()
{
  local title
  title=`_dotfiles_ansi_dequote "$1" | perl -pe "s/;//g"`
  case $TERM in
    xterm*|rxvt*|Eterm|eterm|aixterm|dtterm)
      printf "\033]0;%s\a" "$title"
      ;;
    iris-ansi)
      printf "\033P1.y%s\033\\" "$title"
      ;;
    sun-cmd)
      printf "\033]l%s\033\\" "$title"
      ;;
    hpterm)
      printf "\033&f0k%dD%s" "${#title}" "${title}"
      ;;
    screen*|tmux*)
      # screen and tmux window title
      printf "\033k%s\033\\" "$title"
      # set the tmux pane title too
      [[ -n "$TMUX" ]] && printf "\033]2;%s\033\\" "$title"
      ;;
  esac
}

# Run by zsh immediately before we exec a command; we use it to amend
# the window title to show what is running.
function preexec()
{
  local cmd
  cmd="$1"
  case $cmd in
    fg*)
      # show something more useful than "fg" when we resume a suspended job
      local job
      read cmd job <<< "$cmd"
      if [[ -z "$job" ]] ; then
        # the echo is to strip extra whitespace
        cmd=$(echo $(builtin jobs -l %+ 2>/dev/null | cut -d' ' -f6-))
      else
        cmd=$(echo $(builtin jobs -l $job 2>/dev/null | cut -d' ' -f6-))
      fi
      ;;
  esac
  _dotfiles_set_window_title "$cmd -- ${USER}@${HOSTNAME%%.*}:${PWD/$HOME/~}"
}
```

**Reply by McFrappe** [reply]
That seems to do it good enough for me to be satisfied at least. Thank you so much for this. I will tweak this and see what i can get to work. Once again, thank you @wez !! <3

**Reply by wez** [reply]
You can now do this in `main`:

```lua
local wezterm = require 'wezterm'

wezterm.on("update-right-status", function(window, pane)
  window:set_right_status(pane:get_foreground_process_name())
end)

wezterm.on("format-tab-title", function(tab, tabs, panes, config, hover, max_width)
  local pane = tab.active_pane
  local title = pane.foreground_process_name
  local color = "navy"
  if tab.is_active then
    color = "blue"
  end
  return {
    {Background={Color=color}},
    {Text=" " .. title .. " "},
  }
end)

return {
}
```

This should be available in the nightly download within an hour.

Docs at: https://wezfurlong.org/wezterm/config/lua/PaneInformation.html and https://wezfurlong.org/wezterm/config/lua/pane/get_foreground_process_name.html

**Reply by McFrappe** [reply]
Nice! This was even better. Appreciate the work @wez. 👍

**Reply by McFrappe** [reply]
Altough, the title is very long if a process is deep within a directory since it seems that `pane.foreground_process_name` fetches the path from root.

I tried to write a split function that supposedly should split at each seperator specified:

```lua
local function strip_path_process_name(input)
    local result = {};
    for str in string.gmatch(input, "([^/]*)") do
        table.insert(result, str);
    end
    return result;
end
```

Problem is that this does not seem to do anything when i apply it to for instance this:

```lua
wezterm.on("format-tab-title", function(tab)
    local pane = tab.active_pane
    -- local process = strip_path_process_name(pane.foreground_process_name)
    local title = pane.foreground_process_name
    local color = "navy"

    if tab.is_active then
        color = "red"
    end
    return {
        {Background={Color=color}},
        {Text=" " .. title[0] .. ""},
    }
end)
```
Note that i am just trying to see if the tab changes to anything after applying it through the function, that's why im just picking out the first element in the array. 

Am i missing something on why this does not work as i am trying to get it to?

**Reply by McFrappe** [reply]
<img width="227" alt="Screenshot 2021-12-24 at 5 56 51 PM" src="https://user-images.githubusercontent.com/17787601/147365781-246310ea-bfc9-4254-85c9-9e365fcca491.png">

For reference on what i mean, the active tab that is open is neovim.

**Reply by wez** [reply]
I think you want `basename`; here's an implementation I found in lua with a quick google search: https://github.com/Donearm/scripts/blob/master/lib/basename.lua

**Reply by wez** [reply]
I've added this function to the examples in the docs:

```lua
-- Equivalent to POSIX basename(3)
-- Given "/foo/bar" returns "bar"
-- Given "c:\\foo\\bar" returns "bar"
function basename(s)
  return string.gsub(s, "(.*[/\\])(.*)", "%2")
end
```

**Reply by McFrappe** [reply]
Oh, i see, that did the job. Good that it was added to the docs.

 @wez , thankful for the help getting this available. You've done more than enough.
<img width="227" alt="Screenshot 2021-12-24 at 9 19 51 PM" src="https://user-images.githubusercontent.com/17787601/147371350-baeef350-adcb-45df-8002-171acd00e782.png">
Now it looks nice! 

Merry christmas!

**Reply by Bobrokus** [reply]
I have a different problem.

Because I use pwsh the foreground_process_name always points to pwsh.

I'm currently trying to get around that using a batch script that checks if my code editor process (Helix) is running. If it is, it sets an environment variable. I can then use the env variable in the code. The problem though is that I need to run that script somehow. I was thinking about running it with `os.execute` in `.wezterm.lua` but the function creates a cmd window every time so it's unusable.

I was also thinking about editing the batch script so it runs in a loop. I would then call the command only once and let it run in the background. I need to figure out how I'm going to call it though. If I just put it in the `.wezterm.lua` file it prevents wezterm from starting up. I was thinking about putting it inside an event callback that runs on startup. I didn't get to implementing that yet though.

**Reply by wez** [reply]
@Bobrokus I'd suggest looking at https://wezfurlong.org/wezterm/recipes/passing-data.html for ways to pass data from pwsh to your wezterm lua config.  Presumably there's a general pwsh hook/event that you can use to eg: set a user var to indicate what program is running.

**Reply by Bobrokus** [reply]
@wez I'll have a look. Thanks!

**Reply by jyyhuang** [reply]
@wez Is this the same with wsl? I have the same problem with foreground_process_name pointing to wslhost.exe.

**Reply by geophpherie** [reply]
This is where I'm stuck too. All my tabs say that regardless of where i navigate to. Which, i get because wslhost.exe is running inside of wezterm, but i'm trying to figure out how i can signal out my cwd to update the tab (or if it's even possible).

**Reply by byddha** [reply]
If anyone sees and wants this, this is how I'm doing it for now. I only need a few specific programs so that I can use then for icons.

In my powershell profile:
```pwsh
function Set-WezTermUserVar {
    param($Name, $Value)
    $encoded = [Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($Value))
    $esc = [char]0x1b
    $bel = [char]0x07
    $seq = "${esc}]1337;SetUserVar=${Name}=${encoded}${bel}"
    Write-Host $seq -NoNewline
}

function Invoke-WithOsc1337 {
    param(
        [string]$Command,
        [Parameter(ValueFromRemainingArguments)]
        $Arguments
    )
    Set-WezTermUserVar "PROG" $Command
    try {
        $cmdPath = (Get-Command $Command -CommandType Application | Select-Object -ExpandProperty Source)
        & "$cmdPath" @Arguments
    }
    finally {
        Set-WezTermUserVar "PROG" ""
    }
}

function nvim { Invoke-WithOsc1337 nvim @args }
function lazygit { Invoke-WithOsc1337 lazygit @args }
... etc
```

In wezterm:
```lua
---@diagnostic disable: undefined-field

local wt = require "wezterm"

local M = {}

local current_program = nil

wt.on("user-var-changed", function(_, _, name, value)
  if name == "PROG" then
    current_program = value ~= "" and value or nil
  end
end)

function M.get_process_osc1337()
  return current_program
end

return M

```
